src/hash.c

```diff
@@ -518,6 +518,20 @@ ht_foreach(mrb_state *mrb, htable *t, mrb_hash_foreach_func *func, void *p)
   }
 }
 
+mrb_int
+mrb_os_memsize_of_hash_table(mrb_value obj)
+{
+  struct htable *h = mrb_hash_ptr(obj)->ht;
+  mrb_int segkv_size = 0;
+
+  if(h->index) segkv_size = (sizeof(struct segkv) * h->index->capa);
+
+  return sizeof(htable) +
+    sizeof(segindex) +
+    (sizeof(segment) * h->size) +
+    segkv_size;
+}
+
```

Comment:
* The result seems to be incorrect.

Example (64-bit no-boxing):

| Hash Size  | Expected Result | Actual Result |
|-----------:|----------------:| -------------:|
|          1 |             184 |            72 |
|          8 |             360 |           184 |
|         16 |             600 |           312 |
|         24 |            1176 |          1464 |
|         32 |            1800 |          2616 |

* IV table size should also be included.

* The result may not fit into `mrb_int`.

Reply:
Thank you for this example! Do you mean that a Hash object should also include `mrb_obj_iv_tbl_memsize`, like this

```diff
--- a/mrbgems/mruby-os-memsize/src/memsize.c
+++ b/mrbgems/mruby-os-memsize/src/memsize.c
@@ -74,6 +74,7 @@ os_memsize_of_object(mrb_state* mrb, mrb_value obj)
     }
     case MRB_TT_HASH: {
       size += mrb_objspace_page_slot_size() +
+              mrb_obj_iv_tbl_memsize(mrb, obj) +
               mrb_os_memsize_of_hash_table(obj);
       break;
     }
```

Do you mean the numeric limit of `mrb_int`? That has a minimum limit of `INT32_MAX`. I suppose it's possible for one object consuming 2gb of memory.

Reply:
> Do you mean that a Hash object should also include `mrb_obj_iv_tbl_memsize`, like this

Since IV table is part of a `Hash` object, I think it's natural that its memory size be included in the result of `mrb_os_memsize_of_hash_table()`.

However, in CRuby, for example, the result of `rb_ary_memsize()` seems not to include memory size of IV table, so I think that it will look like the code you presented if follow CRuby.

> Do you mean the numeric limit of mrb_int?

Yes. I have no opinion on the behavior when it exceeds `MRB_INT_MAX`, but I'd like to know if it is not taken into consideration or if it is considered and fixed.